### DISCUSSION TITLE
Is there a quickstart example on how to create a beautiful sequence diagram with ReactFlow?

### DISCUSSION THREAD

**Discussion by victorng8** [markdoc/markdoc#511] [Q&A] [answered]
I was reading the https://stripe.com/blog/markdoc blog there was a snippet of code to generate the sequence diagram like below  

```
{% diagram type="sequence" description="Usage-based billing" %}
  {% node #customer icon="customer" %}
  Customer
  {% /node %}
  {% node #typographic icon="platform" %}
  Typographic
  {% /node %}
  {% node #stripe icon="stripe" %}
  Stripe
  {% /node %}
  {% edge from="customer" to="typographic" %}
  Select plan
  {% /edge %}
  {% edge from="typographic" to="typographic" %}
  Create [usage records](/docs/api/usage_records)
  {% /edge %}
```

however going through the [reactflow](https://reactflow.dev/), I could not find any example of instantiating the sequence diagram

**Comment by mfix-stripe** [comment]
Cc @adreyfus-stripe @mulan-stripe

**Reply by mulan-stripe** [reply]
Hi! You're right, reactflow does not support sequence diagrams out of the box. We ended up creating our own custom implementation using the [custom node](https://reactflow.dev/learn/customization/custom-nodes) abilities in reactflow. This way we were able to create markdoc components that depending on the `type` specified, render different sets of nodes and edges.

You are able to see an example of them [here](https://docs.stripe.com/billing/subscriptions/third-party-payment-processing#collecting-and-storing-customers%E2%80%99-payment-method-details).

**Reply by victorng8** [reply]
Can you please share me the source code for custom node? Like how to create the sequence diagram with color

**Reply by mulan-stripe** [reply, marked as answer]
Unfortunately we aren't able to share our source for the custom node as it contains Stripe-specific code. 

Generally however, we have 2 separate types of nodes: actor and action. Actor nodes are presented by the vertical lines, while action nodes are the text that describes the action happening between two actors. 

To visually create the actor nodes with alternating "lines" and "boxes", the entire actor node is split into sections which are either highlighted or not. The highlighted sections have the "box" which is colored, while the unhighlighted sections are simply a "line". Each Section then has `Handle`'s defined ([react flow doc](https://reactflow.dev/api-reference/components/handle)) that are used to attach the edges.  

The action nodes are fairly straightforward. They're a div with handles on each side of the node and containing the text specified between actors. In the example snippet you shared, the action nodes are representing the `{% edge %}` tag with text. 

I hope that helps a bit!
